sklearn/preprocessing/label.py

```diff
@@ -54,69 +57,117 @@ def _check_numpy_unicode_bug(labels):
     >>> from sklearn.linear_model import LinearRegression
-    >>> from sklearn.preprocessing.label import TargetTransformer
-    >>> tt = TargetTransformer(LinearRegression(), func=np.log,
-    ...                        inverse_func=np.exp)
+    >>> from sklearn.preprocessing.label import TransformedTargetRegressor
+    >>> tt = TransformedTargetRegressor(estimator=LinearRegression(),
+    ...                                 func=np.log, inverse_func=np.exp)
     >>> X = np.arange(4).reshape(-1, 1)
     >>> y = np.exp(2 * X).ravel()
     >>> tt.fit(X, y)
     ... #doctest: +NORMALIZE_WHITESPACE
-    TargetTransformer(estimator=LinearRegression(copy_X=True,
-        fit_intercept=True, n_jobs=1, normalize=False), func=<ufunc 'log'>,
-        inverse_func=<ufunc 'exp'>)
+    TransformedTargetRegressor(check_invertible=True,
+        estimator=LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1,
+                                   normalize=False),
+        func=<ufunc 'log'>, inverse_func=<ufunc 'exp'>, transformer=None)
     >>> tt.score(X, y)
     1.0
     >>> tt.estimator_.coef_
     array([ 2.])
 
-
-    Notes
-    -----
-    It is not checked whether func and inverse_func are actually
-    inverse to each other.
     """
-    def __init__(self, estimator, func, inverse_func):
+    def __init__(self, estimator, transformer=None,
+                 func=None, inverse_func=None, check_invertible=True):
         self.estimator = estimator
+        self.transformer = transformer
         self.func = func
         self.inverse_func = inverse_func
+        self.check_invertible = check_invertible
         # we probably need to change this ones we have tags
         self._estimator_type = estimator._estimator_type
 
+    def _validate_transformer(self, y):
+        if (self.transformer is not None and
+                (self.func is not None or self.inverse_func is not None)):
+            raise ValueError("Both 'transformer' and functions 'func'/"
+                             "'inverse_func' cannot be set at the same time.")
+        elif self.transformer is not None:
+            self.transformer_ = clone(self.transformer)
+        else:
+            self.transformer_ = FunctionTransformer(
+                func=self.func, inverse_func=self.inverse_func, validate=False)
+        self.transformer_.fit(y)
+        # XXX: is it a necessary test? one might not want an invertible
+        # function.
+        if self.check_invertible:
+            n_subsample = min(1000, y.shape[0])
+            subsample_idx = np.random.randint(y.shape[0], size=n_subsample)
+            diff = np.abs((y[subsample_idx] -
+                           self.transformer_.inverse_transform(
+                               self.transformer_.transform(y[subsample_idx]))))
+            if np.sum(diff) > 1e-7:
+                raise ValueError("The provided functions or transformer are"
+                                 " not strictly invertible.")
```

Comment:
hm maybe say the functions are not the inverse of each other? they might be invertible but not be the right functions.